sdk/core/azure-core-http-okhttp/src/main/java/com/azure/core/http/okhttp/OkHttpAsyncHttpClient.java

```diff
@@ -98,37 +114,63 @@ private static Mono<okhttp3.Request> toOkHttpRequest(HttpRequest request) {
      * @param headers the headers associated with the original request
-     * @return the Mono emitting okhttp3.RequestBody
+     * @return the Mono emitting okhttp request
      */
-    private static Mono<RequestBody> toOkHttpRequestBody(Flux<ByteBuffer> bbFlux, HttpHeaders headers) {
-        Mono<okio.ByteString> bsMono = bbFlux == null
-            ? EMPTY_BYTE_STRING_MONO
-            : toByteString(bbFlux);
+    private Mono<RequestBody> toOkHttpRequestBody(BinaryData bodyContent, HttpHeaders headers) {
+        String contentType = headers.getValue("Content-Type");
+        MediaType mediaType = (contentType == null) ? null : MediaType.parse(contentType);
 
-        return bsMono.map(bs -> {
-            String contentType = headers.getValue("Content-Type");
-            MediaType mediaType = (contentType == null) ? null : MediaType.parse(contentType);
+        if (bodyContent == null) {
+            return Mono.just(RequestBody.create(ByteString.EMPTY, mediaType));
+        }
 
-            return RequestBody.create(bs, mediaType);
-        });
+        BinaryDataContent content = BinaryDataHelper.getContent(bodyContent);
+
+        if (content instanceof ByteArrayContent) {
+            return Mono.just(RequestBody.create(content.toBytes(), mediaType));
+        } else if (content instanceof StringContent
+            || content instanceof SerializableContent) {
+            return Mono.fromCallable(() -> RequestBody.create(content.toBytes(), mediaType));
+        } else if (content instanceof InputStreamContent) {
+            // The OkHttpInputStreamRequestBody doesn't read bytes until it's triggered by OkHttp dispatcher.
+            return Mono.just(new OkHttpInputStreamRequestBody((InputStreamContent) content, headers, mediaType));
+        } else if (content instanceof FileContent) {
+            // The FileContent doesn't read bytes until it's triggered by OkHttp dispatcher.
+            return Mono.just(new OkHttpFileRequestBody((FileContent) content, headers, mediaType));
+        } else {
+            // The OkHttpFluxRequestBody doesn't read bytes until it's triggered by OkHttp dispatcher.
+            OkHttpFluxRequestBody fluxRequestBody = new OkHttpFluxRequestBody(
+                content, headers, mediaType, writeTimeout);
+            if (fluxRequestBody.contentLength() < 0
+                || fluxRequestBody.contentLength() > BUFFERED_FLUX_REQUEST_THRESHOLD) {
+                // If Flux is large enough or length is unknown (aka chunked encoding)
+                // then the cost of thread jumping (see OkHttpFluxRequestBody class)
+                // is lesser than memory pressure.
+                // This allows arbitrary length uploads.
+                return Mono.just(fluxRequestBody);
+            } else {
+                // If Flux is small enough then buffering performs better.
+                return toByteString(bodyContent.toFluxByteBuffer()).map(bs -> RequestBody.create(bs, mediaType));
```

Comment:
If this path is removed, could this API become non-reactive?

Reply:
Yes. this is how it was for a moment before I brought it back.